message_edit: Do not move messages older than 3 days.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First commit is the backend work to add `check_time_limit_constraints` parameter and ignore messages older than 3 days.
- Second commit is the frontent work to show the confirmation modal when user is trying to move messages older than 3 days.
 <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![message-move-time-limits](https://user-images.githubusercontent.com/35494118/208736261-8ed7024d-d945-498e-867d-0d6c37dc22bb.gif)


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented on templates/zerver/api/changelog.md:
```diff
@@ -20,6 +20,14 @@ format used by the Zulip server that they are interacting with.
+* [`PATCH /messages/{message_id}`](/api/update-message): Added `check_time_limit_constraints`
+  parameter.
+* [`PATCH /messages/{message_id}`](/api/update-message): Messages older than 3 days are either
+  ignored or server returns error according to the value of `check_time_limit_constraints`
```
I think this can be written in a more clear and readable way. Any suggestions?

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -6561,6 +6566,21 @@ paths:
+            messages that were older than 3 days. Otherwise, the server ignores the
+            messages older than 3 days and moves rest of the messages to the new topic.
+
+            **Changes**: New in Zulip 7.0 (feature level 160).
```
I think the wording here and above can be improved. Would appreciate any suggestions to do so.

sahil839 commented on static/templates/confirm_dialog/confirm_moving_messages.hbs:
```diff
@@ -0,0 +1,19 @@
+	{{#tr}}
+	You can't move all the messages in this topic, but you can move the latest {{messages_allowed_to_move}} messages. Do you want to go ahead and do that? {{messages_not_allowed_to_move}} messages will remain in the current topic.
+	{{/tr}}
+	{{/if}}
```
I think there are too many if-else blocks here, but I am unable to figure out a better way to do it and also use "messages" or "message" correctly when required. Also, we cannot have "if" tags inside "tr" tags.

alya commented:
Hm, let me know if I'm doing something wrong here...

I logged in as Aaron, posted a message on a topic that had some old messages, went into the "Move messages" modal from the new messages, and selected "Move all messages". The result was that the "Confirm" button is disabled; I didn't get a confirmation dialog.

alya commented on static/templates/confirm_dialog/confirm_moving_messages.hbs:
```diff
@@ -0,0 +1,19 @@
+<p>
```
I think there's a better way to handle plurals for internationalization. https://zulip.readthedocs.io/en/latest/translating/internationalization.html#frontend-translations

alya replied:
We should also split this into two separate translation blocks, to reduce duplication for translators.

sahil839 replied:
Updated the code to use `plural` format.

And I am not sure what you meant by "spllit into two separate translation blocks", do we want to tag different sentences for translation separately?

alya replied:
Yeah, the suggestion is to tag the sentence with the first plural separately from the sentence with the second plural, so that each sentence can be translated once.

sahil839 commented:
> Hm, let me know if I'm doing something wrong here...
> 
> I logged in as Aaron, posted a message on a topic that had some old messages, went into the "Move messages" modal from the new messages, and selected "Move all messages". The result was that the "Confirm" button is disabled; I didn't get a confirmation dialog.

The "Confirm" button is disabled until the topic/stream is changed. And just to be clear, in this PR the time limit only applies for topic editing. #23916 adds time limit for changing streams.

alya commented:
Ah, yeah, that helps -- thanks!

Let's tweak the modal a bit:

- Add a `?` at the end of the modal heading
- Add a line break before "N messages will remain in the current topic."

Otherwise looks good to me!

sahil839 commented:
Updated the PR -
- Added `?` at the end of modal heading.
- Added line break before "N messages will remain in the current topic." and also updated the code to have 3 different translation blocks for 3 sentences.

sahil839 commented:
Also, I noticed that the messages count in notification sent when moving the message is not correct, need to fix it.

And currently in this PR, the rule of moving messages that are recent enough applies when moving between streams also, even when there is no limit on moving messages between streams. I think completing #23916 first would help us in avoiding complicated changes in this PR.

timabbott commented:
Let's try to get this in shape fairly quickly, since the intermediate state after #23916, while coherent, is a lot less nice than the once with this finished as well.

timabbott commented on zerver/actions/message_edit.py:
```diff
@@ -961,6 +967,22 @@ def check_update_message(
             raise JsonableError(_("The time limit for editing this message's topic has passed"))
 
+        if check_time_limit_constraints and propagate_mode == "change_all":
+            message_query = Q(
```
Let's call this something like `disallowed_messages_query`

sahil839 replied:
Udpated this to `allowed_messages_query` since I have changed the logic now.

timabbott commented on zerver/lib/exceptions.py:
```diff
@@ -473,3 +474,25 @@ class ValidationFailureError(JsonableError):
+    def msg_format() -> str:
+        return _(
+            "Can't move all messages in this topic. You can only move {messages_allowed_to_move}/{total_messages_in_topic} latest messages"
+        )
```
I think I like the idea of this explaining which restriction (stream move or topic move) is the issue. Also this should mention "permission". For example: "You do not have permission to move messages older than {date} between streams. Onl {messages_allowed_to_move}/{total_messages_in_topic} will be moved."

sahil839 replied:
Are we fine with showing date when the time limit is set to some minutes or hours. We can though mention that error is due to time limit to improve the message.

Also, there can be a case when both stream and topic move is requested and both are not allowed for some messages and we move the messages for which both are allowed. So, the number of messages not allowed to do stream and topic edit might be different and also the date after which messages are allowed to move would be different. Do, we need to mention both of them in such case?

timabbott replied:
Yeah let's go with a bit simpler. Perhaps `You only have permission to move the most recent messages in this topic ({total_messages_allowed_to_move}/{total_messages_in_topic}).`

timabbott commented on zerver/views/message_edit.py:
```diff
@@ -125,6 +125,7 @@ def update_message_backend(
     send_notification_to_old_thread: bool = REQ(default=False, json_validator=check_bool),
     send_notification_to_new_thread: bool = REQ(default=True, json_validator=check_bool),
     content: Optional[str] = REQ(default=None),
+    check_time_limit_constraints: bool = REQ(default=False, json_validator=check_bool),
```
You should start a thread in "api design" about how we want this API to work. I think the behavior we want long-term is that this endpoint, by default, will throw the new exception if you request a move that you don't have permission to fully complete -- a client is unlikely to actually want to do that. 

I think making partial moves that seem like they're not intended throw an error is a higher priority for me that avoiding backwards-compatibility, so I think we probably want something like `partial_move_if_limited` if we want a parameter at all. I think the other option would be for the exception to include the message ID of the first message that you do have permission to move, and a client that hits the "confirm" button should send a new request that's `change_later` with that message ID as the anchor, with no new API parameter.

sahil839 commented:
Updated the PR.
- Added a commit to not treat "no topic" messages as a special case as we discussed. Note that the "pencil" icon in recipient bar is still highlighted always without hovering for "no topic" messages.
- Updated the last two commits to follow the new approach discussed in https://chat.zulip.org/#narrow/stream/378-api-design/topic/partially.20move.20messages.20as.20per.20time.20limit.20setting

alya commented:
Sorry, I didn't see that this was updated. Works for me in manual testing!

Let's add some vertical space before the last line in the confirmation modal -- I think that will look nicer.

![Screen Shot 2023-03-23 at 2 32 13 PM](https://user-images.githubusercontent.com/2090066/227368945-eaadcc33-4537-43c6-bc06-31147e9bd685.png)

alya commented:
When testing a different PR, I noticed that we aren't handling the case where a user is allowed to move messages to a different stream but not to a different topic correctly. The topic field in the "Move topic" modal should be disabled, and it's not:

![Screen Shot 2023-03-23 at 3 57 45 PM](https://user-images.githubusercontent.com/2090066/227384933-19000020-0cbf-4363-980d-cc91ee3d7907.png)

alya commented:
I don't know if this PR happens to fix this?

sahil839 commented:
Updated the PR.
- Added space before last line in the modal.
- Added commit to disable the topic input if required as per `edit_topic_policy`.

timabbott commented on zerver/actions/message_edit.py:
```diff
@@ -885,6 +886,77 @@ def user_info(um: UserMessage) -> Dict[str, Any]:
+        first_message_allowed_to_move=messages_allowed_to_move[0],
+        total_messages_in_topic=len(all_messages_in_topic),
+        messages_allowed_to_move=len(messages_allowed_to_move),
+    )
```
I have a few concerns with the logic here. One is that you're doing direct queries on `Message`, which don't check whether the acting user has permission to see all of those messages -- a potential security concern in a stream without shared history. 

The second is that we do a very similar set of queries inside `do_update_message`, being careful about this security issue, to determine what sort of notice to display from Notification Bot on whether the entire topic was moved or not; and probably it'd be good to avoid duplicate code/queries for that, which I suspect (but have not confirmed) we're doing here.

If we didn't need to think about the permissions issue, we could do this query much more efficiently with just filtering `messages_for_topic` for the oldest message in the topic, and checking whether or not it's older? It's possible we want to do that check to have a fast path for the case that it's going to be allowed, since that query is likely very fast, and fetching all the message IDs in a topic may not be?

Idunno, maybe that's too complicated, and we should just have two cases -- (1) the current user has full history access, where we can roughly what you have above, and (2) they don't, and we need to compute both `messages_allowed_to_move` and `all_messages_in_topic` via a query that filters on `UserMessage` too. 

(In any case, it was probably a good idea you put this in its own function given how complicated it's looking like it'll be!)

timabbott replied:
> I am not sure whether we can just deduplicate code though by extracting some functions, but I think it would make code somewhat hard to read. In any case, I can try if we can do this.

Yeah, maybe it's fine to leave this as-is.

sahil839 commented on zerver/tests/test_message_edit.py:
```diff
@@ -1353,9 +1353,9 @@ def assert_is_topic_muted(
 
         change_all_topic_name = "Topic 1 edited"
-        # This code path adds 17 (10 + 4/visibility_policy + 1/user + 1) to
+        # This code path adds 20 (13 + 4/visibility_policy + 1/user + 1) to
```
To be honest, I think this comment is not really undestandable. I am not sure what "This code path" means here and what queries does it add to moving topic code. I think all the database work for moving a topic happens in `check_update_message` which is tested here and this itself is the function which is called from the views function. 

I remembered trying to look at the commit history for this line, but did not came to a conclusion then. Would try to look at it once again.

timabbott replied:
I think either way works for me.

timabbott commented:
Can you post in "api design" the precise set of parameters in the new API response? I'd like to make sure we get feedback on that detail, and am not sure we've done that.

timabbott commented:
I think it's possible you could optimize the queries at the end of `check_time_limit_for_change_all_propagate_mode` for the public stream case by just fetching the set of `(message_id, date_sent)` values in the topic and then doing the date filtering in Python -- I believe the current construction goes to the database twice.

sahil839 commented:
Updated the PR -
- Modified `check_time_limit_for_change_all_propagate_mode` to reduce the number of queries for public stream case.
- Added commit to modify the test comment. The commen may require rephrasing though.
- Updated the error response variable names and added the API documentation.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -6622,25 +6633,39 @@ paths:
+                        description: |
+                          A typical JSON response for when one doesn't have the permission to
+                          edit a particular message:
+                  - allOf:
+                      - $ref: "#/components/schemas/CodedError"
+                      - example:
+                          {
+                            "code": "MESSAGE_MOVE_TIME_LIMIT_EXCEEDED",
```
Agreed that we should do this, though I'm OK with it being a follow-up just to get all the other complexity of this change into `main`.

timabbott replied:
I ended up working on this despite my previous comment, here's the diff:

``` diff
                       - example:
                           {
                             "code": "MESSAGE_MOVE_TIME_LIMIT_EXCEEDED",
+                            "first_message_id_allowed_to_move": 123,
                             "msg": "Can't move all messages in this topic. You can only move 2/5 latest messages",
                             "result": "error",
+                            "total_messages_allowed_to_move": 2,
+                            "total_messages_in_topic": 5,
                           }
                         description: |
-                          A typical failed JSON response for when time limit for moving messages has
-                          passed.
+                          A special failed JSON response for when the user has permission to move
+                          the target message, but asked to move all messages in a topic
+                          (`change_all`) and the user does not have permission to move the entire
+                          topic.
+
+                          This happens when the topic contains some messages that are older than an
+                          applicable time limit for the requested topic move
+                          (`move_messages_within_stream_limit_seconds` and/or
+                          `move_messages_between_streams_limit_seconds`).
+
+                          This response contains data on which portion of this topic the user has
+                          permission to move; `first_message_id_allowed_to_move` is the oldest
+                          message ID in this topic that the user has permission to move; the user
+                          can move `total_messages_allowed_to_move` messages, and
+                          `total_messages_in_topic` are outside the user's permissions.
+
+                          A client is recommended to either just present the error to the user, or
+                          prompt the user with this information. In such a prompt, clients should
+                          offer the option of cancelling and (manually) asking a moderator to move
+                          the entire topic, since that's often a better experience than partially
+                          moving a topic. This API supports a client that wishes to allow the user
+                          to repeat the request with a `change_later` propagation mode and a target
+                          message ID of `first_message_id_allowed_to_move`, if desired.
+
+                          Note that in a stream with protected history, the Zulip security model
+                          requires that the above calculations only include messages the acting user
+                          has access to.
 
                           **Changes**: New in Zulip 7.0 (feature level 172).
     delete:
```

timabbott commented on zerver/actions/message_edit.py:
```diff
@@ -930,6 +930,98 @@ def user_info(um: UserMessage) -> Dict[str, Any]:
+        )
+
+    # If the code reaches here, then there is a time limit to edit at least one of topic or stream.
+    assert topic_edit_deadline_seconds is not None or stream_edit_deadline_seconds is not None
```
This assertion is I wrong, in that you could have a time limit only on moving messages between streams, but this specific request is only to move the topic within the stream. I fixed this by just making the `return` case an `else` clause a bit further down, which also feels a lot more readable.

As a side note, I line-wrapped a bunch of your comments; generally we want to line-wrap comments to ~72 characters, unless they contain a URL or something that provides a reason not to do so.

sahil839 replied:
Thanks for catching that.

timabbott commented:
Updated this for the above two details. We likely will want to rewrite the text/modal given the context I added to the API documentation, that probably in most cases we want to plant the idea of asking a moderator to do the move instead of going forward as the default action.

sahil839 commented:
We can probably add a line like "You can also ask a moderator to move the complete topic instead." after "Do you want to go ahead and do that?". But we would have the text about how many messages will remain in the topic after that. So, we might want to add the text about asking moderators at the end in a new line.

Current modal -
![Screenshot from 2023-04-05 08-40-56](https://user-images.githubusercontent.com/35494118/229972053-7d377282-faf4-4f32-8d69-fe548d45ba47.png)

alya commented:
How about something like this for the modal text?

> You do not have permission to move some of the messages in this topic. Contact a moderator to move all messages.

> Do you still want to move the latest 2 messages? 3 messages will remain in the current topic.

alya commented:
Draft copy for modal to show when a user attempts to resolve a topic where some of the messages cannot be moved:

## Could not resolve topic

You only have permission to resolve newer topics in this organization. Contact a moderator to resolve this topic.

sahil839 commented:
> How about something like this for the modal text?
> 
> > You do not have permission to move some of the messages in this topic. Contact a moderator to move all messages.
> 
> > Do you still want to move the latest 2 messages? 3 messages will remain in the current topic.

Looks good to me. Will update the PR.

timabbott commented:
> Can we do the resolve topic changes in a follow-up PR?

Yeah I think it's probably fine to do that, but we should aim to do those changes this week.

sahil839 commented:
I missed one question in the previous comment. Do we want the text in first para shown in the modal to be something different when there is only one message which user cannot move or "do not have the permission to move some of the messages" is fine.

sahil839 commented:
Updated the error message returned by API.

> Yeah I think it's probably fine to do that, but we should aim to do those changes this week.

I guess this could be completed this week. Have started discussion in https://chat.zulip.org/#narrow/stream/378-api-design/topic/partially.20move.20messages.20as.20per.20time.20limit.20setting

gnprice commented:
For cross-reference: this appears to be related to #21739.
